lib/UR/Role.pm

```diff
@@ -0,0 +1,732 @@
+definition.
+
+  class My::Class {
+      roles => ['My::Role', 'Other::Role'],
+      is => ['Parent::Class'],
+      has => ['prop_a','prop_b'],
+  };
+
+Properties and meta-attributes from the Roles get copied into the composing
+class.  Subroutines defined in the Roles' namespaces are imported into the
+class's namespace.  Operator overloads defined in the Roles are applied to
+the class.
+
+=head3 Property and meta-attribute conflicts
+
+An exception is thrown if multiple Roles are composed together that define
+the same property, even if the composing class defines the same property in
+an attempt to override them.
+
+=head3 Method conflicts
+
+An exception is thrown if multiple Roles are composed together that
+define the same subroutine name unless the composing class's namespace also
+has a subroutine with the same name.  The class is expected to implement
+whatever behavior is required, maybe by calling one role's method or the
+other, both methods, neither, or anything else.
+
+To call a function in a role, the function's fully qualified name, including
+the role's package, must be used.
+
+=head3 Overload conflicts
+
+Like with method conflicts, an exception is thrown if multiple Roles are
+composed together that overload the same operator unless the composing
+class also overloads that same operator.
+
+An exception is also thrown if composed roles define incompatible 'fallback'
+behavior.  If a role does not specify 'fallback', or explicity sets it to
+C<undef>, it is compatible with other values.  A Role that sets its 'fallback'
+value to true or false is only compatible with other roles' values of undef
+or the same true or false value.
+
+=head2 __import__
+
+Each time a Role is composed into a class, its C<__import__()> method is
+called.  C<__import__()> is passed two arguments:
+
+=over 4
+
+=item *
+
+The name of the role
+
+=item *
+
+The class metadata object composing the role.
+
+=back
+
+This happens after the class is completely constructed.
+
+=head2 Deferred Values
+
+A Role definition may contain L<UR::Role::DeferredValue> objects to act as
+placeholders for values to be filled in when the role is composed into a
+class.  These values are resolved at composition time by calling the named
+function on the composing class.  For example:
+
+  use UR::Role;
+  role ObjectDisplayer {
+      has => [
+          target_object => { is => defer 'target_type' },
+      ]
+  };
+
+  class ShowCars {
+      roles => ['ObjectDisplayer'],
+  };
+  sub ShowCars::target_type { 'Car' }
+
+When the 'target_object' property is composed into the ShowCars class, the
+system calls the method C<ShowCars-E<gt>target_type()> to obtain the value
+'Car' for the data_type of property 'target_object'.
+
+UR::Role exports the function C<defer> to create these DeferredValue objects.
+
+=head1 SEE ALSO
+
+L<UR>, L<UR::Object::Type::Initializer>, L<UR::Role::DeferredValue>
+
+=cut
```

Comment:
I don't think we need to worry about it now but the lack of `before`, `after`, and `around` might be limiting.  I think it can be worked around manually by overriding the method in the concrete class and delegating.